main/server/init.js

```diff
@@ -0,0 +1,42 @@
+import { fork } from 'child_process'
+import { join } from 'path'
+import log from 'electron-log'
+
+const START_PORT = 8000
+const MAX_ATTEMPTS = 10
+
+export const startServer = () => {
+  let attempts = 0
+
+  function attemptAStart() {
+    if (attempts >= MAX_ATTEMPTS) {
+      return Promise.reject(`Failed to bind socket server after ${MAX_ATTEMPTS} attempts.`)
+    }
+
+    return new Promise((resolve, reject) => {
+      const randomPort = START_PORT + Math.floor(1000 * Math.random())
+      log.info(`Starting socket server on port: ${randomPort}`)
+
+      const server = fork(join(__dirname, './socketServer.bundle.js'), [randomPort])
+      server.on('close', (code) => {
+        log.warn(`Socket server died with code: ${code}`)
+        if (code === 1) {
+          log.warn(`Restarting the server on a new port.`)
+          attempts++
+          return attemptAStart()
+        }
+        log.error(`Failed with an unhandled error.  Killing the server.`)
+        return reject(new Error(`Socket worker died with unhandled error code: ${code}`))
+      })
+      server.on('message', (message) => {
```

Comment:
Am I reading this correctly that the server isn't "up and running" and considered ready until it receives its first message from the socket worker (I'm assuming that's in a project worker)? Is that going to cause any problems with the timing of different pieces at startup and was that intentional?

Reply:
And I think i missed where in the code it sends the 'ready' message after startup

Reply:
So yes, it considers this as part of the critical path of booting up the app.
We don't need a project open, though.  Here, the main process waits for the forked process to tell it that it started the socket server.

See: https://github.com/cameronsutter/plottr_electron/blob/staging/main/server/server.js#L58
That line uses interprocess communication rather than a socket to let the parent process know that the socket server is up and running on the child process.